WordPress/gutenberg
# Full Site Editing: Add Support for Templates Default and Custom Titles and Descriptions

## Description


Fixes #25755 

Update the templates handling to allow for custom titles and descriptions.

- Created a `theme-provided` (the name is bad, I'm open to suggestions!) custom status for `wp_template` CPT to indicate those templates that have been imported from files and not touched by the user. Previously we were using `auto-draft`, which wasn't super appropriate, and created a few complications (e.g. autodrafts can't have titles, aren't supposed to show up anywhere, etc.).
- Added support for `excerpts` to the `wp_template` CPT. Right now titles and excerpts are only editable when editing the template in its own post editor.
- Added a `theme` post meta, containing the source theme for `theme-provided` templates.
- Updated the Templates wp-admin list to display more details (title, slug, excerpt, status, source), so that now it's at least slightly useful. It also now lists `theme-provided` and (at least temporarily) `auto-draft` templates.
- Created a list of default template types definitions in PHP, containing titles and descriptions of all the "generic" (not containing variables in their slugs) templates.
- The list is filterable with the new `template_types_definitions` filter, which is already used to temporarily remove the `embed` definition, since it's currently not supported.
- The list is available in JS via Redux: it's in `settings.defaultTemplateTypesDefinitions` of `core/edit-site`.
- Templates imported from files will automatically obtain the title and excerpt from the default definitions, if possible.
- New templates created from the Site Editor sidebar will also automatically use the default definitions, and will be saved as `draft`. They will become `publish`ed once saved the first time.
- The default definitions are used across the Site Editor as a fallback for templates without title or excerpt.

### Follow up

- Update the template definitions copy!
- Add a way to edit the title (and maybe excerpt too) of a template from the Site Editor (e.g. in the Document Settings dropdown).
- Extend the wp-admin Templates list's quick edit with the excerpt (and maybe the slug too?).
- Update the template saving logic to allow for going back and forth between `draft` and `publish`.

## How has this been tested?

Note: this will create new templates which are only accessible with the changes included in this PR. Changing branch would make the new templates inaccessible; they shouldn't be used by the site, but are still there, and may cause unexpected conflicts with pre-existing `auto-draft` templates.

1. Open the Site Editor. Behind the scenes, all theme's and plugins' templates will be converted into new `wp_template` items of status `theme-provided`; this means that there will be duplicates.
2. Open the wp-admin Templates list (`/edit.php?post_type=wp_template`).
3. Make sure it shows the following columns: title, slug, description, status.
4. Notice it shows also `auto-draft` templates. Feel free to wipe this list clean by moving all templates to trash. New `theme-provided` templates will be generated again once opening the Site Editor.
5. Open the Site Editor (if you wiped the templates, now it will generate new ones, hopefully without duplicates).
6. Open the sidebar, and navigate into the Templates menu.
7. Make sure the items use the title and description as provided by the definitions list introduced in this PR.
8. Create a new template with the New Template dropdown (click on `+` in the Templates menu).
9. Make sure the items in the New Template dropdown use the definitions list.
10. Make sure the new template's name is the one provided by the definitions list (it should match the one from the New Template dropdown).
11. Check in the sidebar that the new template shows up (it might be easier to just look in the All Templates menu), it has the correct title and description, and is marked as "[Draft]".
12. Click on "Update Design" (it might be disabled: do some changes to enable it).
13. The template has now been published, so make sure it lost the "[Draft]" label.
14. Open the wp-admin Templates list, and make sure the new template shows up with the right title, description, and status.

## Screenshots

The new Templates wp-admin list:

<img width="1185" alt="Screenshot 2020-10-28 at 12 01 10" src="https://user-images.githubusercontent.com/2070010/97433370-4d7d5880-1915-11eb-9805-6bf2d6991ae8.png">

The Site Editor sidebar using the new default definitions, and including a draft template:

<img width="522" alt="Screenshot 2020-10-26 at 17 44 59" src="https://user-images.githubusercontent.com/2070010/97208552-29a20180-17b3-11eb-892b-753f84693ffe.png">

The Site Editor documents settings using the new default definitions:

<img width="276" alt="Screenshot 2020-10-26 at 17 44 41" src="https://user-images.githubusercontent.com/2070010/97208602-3888b400-17b3-11eb-9848-9e6662943acc.png">

## Types of changes
Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
